internal/cloud/state.go

```diff
@@ -33,69 +63,371 @@ of authorization and therefore this error can usually be fixed by upgrading the
+	return statemgr.SnapshotMeta{
+		Lineage: s.lineage,
+		Serial:  s.serial,
+	}
 }
 
-// Unlock delegates calls to unlock state to the remote State
-func (s *State) Unlock(id string) error {
-	return s.delegate.Unlock(id)
+// statemgr.Writer impl.
+func (s *State) WriteState(state *states.State) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// We create a deep copy of the state here, because the caller also has
+	// a reference to the given object and can potentially go on to mutate
+	// it after we return, but we want the snapshot at this point in time.
+	s.state = state.DeepCopy()
+
+	return nil
+}
+
+// statemgr.Persister impl.
+func (s *State) PersistState(schemas *terraform.Schemas) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.readState != nil {
+		lineageUnchanged := s.readLineage != "" && s.lineage == s.readLineage
+		serialUnchanged := s.readSerial != 0 && s.serial == s.readSerial
+		stateUnchanged := statefile.StatesMarshalEqual(s.state, s.readState)
+		if stateUnchanged && lineageUnchanged && serialUnchanged {
+			// If the state, lineage or serial haven't changed at all then we have nothing to do.
+			return nil
+		}
+		s.serial++
+	} else {
+		// We might be writing a new state altogether, but before we do that
+		// we'll check to make sure there isn't already a snapshot present
+		// that we ought to be updating.
+		err := s.refreshState()
+		if err != nil {
+			return fmt.Errorf("failed checking for existing remote state: %s", err)
+		}
+		if s.lineage == "" { // indicates that no state snapshot is present yet
+			lineage, err := uuid.GenerateUUID()
+			if err != nil {
+				return fmt.Errorf("failed to generate initial lineage: %v", err)
+			}
+			s.lineage = lineage
+			s.serial = 0
+		}
+	}
+
+	f := statefile.New(s.state, s.lineage, s.serial)
+
+	var buf bytes.Buffer
+	err := statefile.Write(f, &buf)
+	if err != nil {
+		return err
+	}
+
+	var jsonState []byte
+	if schemas != nil {
+		jsonState, err = jsonstate.Marshal(f, schemas)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = s.uploadState(s.lineage, s.serial, s.forcePush, buf.Bytes(), jsonState)
+	if err != nil {
+		s.stateUploadErr = true
+		return fmt.Errorf("error uploading state: %w", err)
+	}
+	// After we've successfully persisted, what we just wrote is our new
+	// reference state until someone calls RefreshState again.
+	// We've potentially overwritten (via force) the state, lineage
+	// and / or serial (and serial was incremented) so we copy over all
+	// three fields so everything matches the new state and a subsequent
+	// operation would correctly detect no changes to the lineage, serial or state.
+	s.readState = s.state.DeepCopy()
+	s.readLineage = s.lineage
+	s.readSerial = s.serial
+	return nil
+}
+
+func (s *State) uploadState(lineage string, serial uint64, isForcePush bool, state, jsonState []byte) error {
+	ctx := context.Background()
+
+	// Read the raw state into a Terraform state.
```

Comment:
@brandonc I added this in to fix the crash that I was getting on the cloud test. It seems as though `JSONStateOutputs` needs to be populated here, so I just copied what was in [backend_state.go](https://github.com/hashicorp/terraform/blob/5da30c2b65265c9c7ac7580a1295e87715ebb568/internal/cloud/backend_state.go#L66-L79) to do that.

* Is this the correct thing to do and is this the right spot? Wasn't sure if I was supposed to be putting this logic in `PersistState` instead - maybe it doesn't matter?
* Also, do I need to add `JSONState` in `backend_state.go`?

I think this also somehow/some way ties into why I needed to comment out [this test](https://github.com/hashicorp/terraform/pull/31698/files#diff-028e6f10c867390b9523c208ae67c17a730615e3a1e54600cae030c810ecedabR23-R64), but wasn't sure how to fix that.

Reply:
Hmmm I think backend_state.go and state.go are doing the same thing and only one needs to exist. I think it's probably best to delete one altogether and move your changes to the other.

backend_state.go was added in the original PR that was not merged, but after that I added the simpler state.go which was just designed to wrap state persistence in order to save the json outputs while backend_state was designed to be a complete state interface (without an underlying delegate)

I'm not even sure which one is better, but state.go is simpler and already merged. If it can work for you, I would use it.